### ISSUE TITLE
Add support for Nuvision Solo 10 Draw

### ISSUE THREAD

**Issue by ShapeShifter499** [linuxwacom/libwacom#70]
I added my request to https://github.com/linuxwacom/wacom-hid-descriptors/issues/48   But I think it's also valid here.

I'm just not 100% sure how to create the required tablet file.

**Comment by whot** [comment]
See [the wiki](https://github.com/linuxwacom/libwacom/wiki/Tablet-Definition-Files) or just browse through the files in data/ to have a guess. This thing doesn't have any buttons, so it'll likely look quite similar to #58.

**Comment by ShapeShifter499** [comment]
@whot I added a generic "nuvision.tablet" file with the contents of https://github.com/linuxwacom/libwacom/blob/master/data/elan-24db.tablet  I only changed the device ID to '04f3:22f7' and the name to 'nuvision'

It appears to work. It even shows up in Gnome Wacom settings as "nuvision" but I have a new issue.  I can't get the eraser or the side button working from my surface 4 pen. I can use this pen to draw in Xournal and "Stylus Labs Write" (http://www.styluslabs.com/)      But I cannot get the eraser to work. 

When I had Windows I was able to erase things by flipping the pen and making a erasing action with the other end of the pen. 

Do you have any ideas how I can debug for that?

**Comment by whot** [comment]
Generally: device functionality isn't part of libwacom, libwacom just describes the tablet for information that we can't get out of the kernel. If the eraser doesn't work, it's likely that you need a kernel patch (which would also help giving it a decent device name...)

What's the evemu-record output when you use the eraser?

**Comment by ShapeShifter499** [comment]
@whot  I found out that if I hover the pen just to the point where the tablet picks it up I can trigger the side button. Seems to work the best if I hold that button down just as I come in a touch the tablet with the pen. But because the way my hands are (physical condition) I'd rather avoid having to use the side button for any sort of erasing. That side button takes a rather stiff click to activate it. 

With that bit of info above in mind I started up evemu-record and did the following. 

Drew a straight line in Xournal, attempted to erase that line by flipping the pen, and then tried to get the side button to activate.

Attached is the output from evemu-record
[evemu.log](https://github.com/linuxwacom/libwacom/files/2858443/evemu.log)

**Comment by ShapeShifter499** [comment]
@whot and here is another log where all I did was attempt to erase lines I already made in Xournal.
[evemu-erase.log](https://github.com/linuxwacom/libwacom/files/2858506/evemu-erase.log)

**Comment by whot** [comment]
You'll need a kernel patch for this device, but there may be a quirk available, cc @bentiss 

Short story: rubber/pen are supposed to be exclusive but your device sets pen, then rubber instead of clearing the pen bit first. That's not something the current code handles because it's not correct protocol. The `BTN_TOOL_*` bits are special.

```
E: 0.078665 0001 0140 0001	# EV_KEY / BTN_TOOL_PEN         1
E: 0.078665 0003 0000 7086	# EV_ABS / ABS_X                7086
E: 0.078665 0003 0001 4153	# EV_ABS / ABS_Y                4153
E: 0.078665 0000 0000 0000	# ------------ SYN_REPORT (0) ---------- +8ms
E: 0.084891 0001 0141 0001	# EV_KEY / BTN_TOOL_RUBBER      1
E: 0.084891 0004 0004 852037	# EV_MSC / MSC_SCAN             852037
E: 0.084891 0001 014a 0001	# EV_KEY / BTN_TOUCH            1
E: 0.084891 0003 0000 7077	# EV_ABS / ABS_X                7077
E: 0.084891 0003 0001 4176	# EV_ABS / ABS_Y                4176
E: 0.084891 0000 0000 0000	# ------------ SYN_REPORT (0) ---------- +6ms
```

**Comment by ShapeShifter499** [comment]
> Short story: rubber/pen are supposed to be exclusive but your device sets pen, then rubber instead of clearing the pen bit first.

@whot My device?  Is it the pen or the tablet?    Getting a different pen would not solve this issue?

**Comment by whot** [comment]
it's the device, the pen won't make a difference here, I suspect

**Comment by ShapeShifter499** [comment]
@whot ah hmm... I'll wait for @bentiss to mention about the quirks then.

**Comment by ShapeShifter499** [comment]
@whot @bentiss any updates on this?

**Comment by ShapeShifter499** [comment]
Also should "Touch" be set to "true" or "false"?

Touch=true

**Comment by ShapeShifter499** [comment]
@whot does it make any difference that I did not set anything for my pen in xorg config?

**Comment by ShapeShifter499** [comment]
I tried looking through code but I didn't see any quirk that would work with my particular situation. @bentiss @whot

**Comment by ShapeShifter499** [comment]
I wonder if this could be solved by adding a quirk that would clear the pen bit whenever "BTN_TOOL_RUBBER" is detected.

**Comment by ShapeShifter499** [comment]
I tried looking through code but I didn't see any quirk that would work with my particular situation. @bentiss @whot @jigpu

**Comment by ShapeShifter499** [comment]
I wonder if this could be solved by adding a quirk that would clear the pen bit whenever "BTN_TOOL_RUBBER" is detected.

**Comment by ShapeShifter499** [comment]
@whot @jigpu @bentiss There are not a lot of people officially working on Wacom device support for Linux is there?    I didn't realize this till I looked at the "about" page of the Linux Wacom website. https://linuxwacom.github.io

**Comment by whot** [comment]
Well, Wacom has 3 Linux engineers but they obviously work on Wacom tablets, not others (the Nuvision isn't a Wacom-branded tablet). And everyone else is ETIME on everything, especially for "random" hardware, sorry.

The closest thing I can recommend here is to check if the DIGImend project had some 
You could try to see if the DIGImend project has/had some quirks/hints for handling your tablet - the input-wacom tree is the Wacom-branded specific backports and not really the target for your tablet.

Based on the sysinfo from https://github.com/linuxwacom/wacom-hid-descriptors/issues/48, looks like the tablet is handled by the `hid-generic` driver, nod the wacom driver. So in the end, the quirk needs to go into `hid-core.c` in the kernel.


`TOUCH=true` should be set if this thing has a touch device, yes. Do also set the `PairedID` if it's different.

**Comment by bentiss** [comment]
Actually, the driver bound to the device is `hid-multitouch`. It handles the touch part but just relay the pen part to hid-input.c. @whot was not entirely wrong, but not entirely correct (in that situation, `hid-multitouch` doesn't do much more that `hid-generic` for the pen).

That also means that DIGImend might not be able to help. You have a "Microsoft Pen Protocol" tablet, and it should be working out of the box (but it's not).

Could you provide us some hid-recordings of your tablet (please use the python version in https://gitlab.freedesktop.org/libevdev/hid-tools) when:
- approaching the pen in proximity, and removing it (no touch)
- approaching the pen, touch with the tip, removing it
- approaching the rubber side of the pen in proximity, remove it (no touch)
- approaching the rubber side, touch, remove
- while in hovering, click on the side button
- while touching with the normal tip, click on the side button

Please separate all of these in different logs, so we can easily understand which is which.

**Comment by ShapeShifter499** [comment]
@bentiss Attached are the logs generated with "hid-record" from https://gitlab.freedesktop.org/libevdev/hid-tools 
I hope these help.

[pen_prox-no-touch.log](https://github.com/linuxwacom/libwacom/files/2909442/pen_prox-no-touch.log)
[pen_touch.log](https://github.com/linuxwacom/libwacom/files/2909443/pen_touch.log)
[rubber_prox-no-touch.log](https://github.com/linuxwacom/libwacom/files/2909444/rubber_prox-no-touch.log)
[rubber_touch.log](https://github.com/linuxwacom/libwacom/files/2909445/rubber_touch.log)
[side_button-hover-no-touch.log](https://github.com/linuxwacom/libwacom/files/2909446/side_button-hover-no-touch.log)
[side_button-pen-touch.log](https://github.com/linuxwacom/libwacom/files/2909447/side_button-pen-touch.log)

**Comment by bentiss** [comment]
thanks for the logs. This helps, but there are some inconsistencies. Could you record once again `rubber_touch.log`. The `rubber_prox-no-touch.log` makes sense as the `invert` usage is set to 1, but in the touch version, it is not.

**Comment by bentiss** [comment]
BTW, to give you a little bit of background, this is exactly what is sent through the wire from the device before kernel processing. The lines starting with `#` are the human version of it based on the first 328 lines that are the protocol description of the device.

To sum up, if, when you normally press the side button, the logs do not show `Barrel Switch: 1` that means that the kernel is not seeing the physical button press, and that there is nothing we can do to fix it.

**Comment by ShapeShifter499** [comment]
@bentiss Attached are some more logs. Because the side button is hard to press, I'll probably avoid that button all together. I just want this pen to function like a digital pencil. Flipping the pen to erase is much easier for me then trying to click that side button.

[rubber-quick-touch.log](https://github.com/linuxwacom/libwacom/files/2909978/rubber-quick-touch.log) I tapped the screen with the eraser quickly.

[rubber-long-touch.log](https://github.com/linuxwacom/libwacom/files/2909979/rubber-long-touch.log) I touched and dragged the eraser around the screen for a few moments.

[rubber_prox-quick.log](https://github.com/linuxwacom/libwacom/files/2909981/rubber_prox-quick.log) Quickly got the tablet to recognize the eraser but I didn't touch the screen.

[rubber_prox-long.log](https://github.com/linuxwacom/libwacom/files/2909983/rubber_prox-long.log) I tried to get the tablet to recognize the eraser end for a longer period with out touching but I may have touched the screen for small taps by mistake.

**Comment by bentiss** [comment]
Thanks for the extra logs.

So, summary is: it's a mess.

The way it is supported to be reported to the host is described in https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/windows-pen-states#windows-pen-state-transitions

But here, the kernel and this device don't follow it entirely:
- as long as the firmware has detected which side is presented, it reports:
   `In Range: 1 | Tip Switch: 1 | Barrel Switch: 0 | Invert: 0 | Eraser: 0` which is fine
- going from `Pen is in range` to `Pen is in range with the intend to erase`, there is a single report with:
  `In Range: 0 | Tip Switch: 0 | Barrel Switch: 0 | Invert: 0 | Eraser: 0` -> still good
- then we receive a bunch of:
  `In Range: 1 | Tip Switch: 0 | Barrel Switch: 0 | Invert: 1 | Eraser: 0` as long as you keep the pen close enough to the screen where the sensor can detect which end you are using.

This part is correct, but the kernel implementation is not: the Linux implementation expects the `Invert` usage to be sent *before* the `In Range` usage -> this explains why `BTN_TOOL_PEN` is not cleared.
This is a bug, but it's non trivial to fix unfortunately.

Then comes the faulty firmware:
- Moving from `Pen is in range with the intend to erase` to `Pen is erasing`, we receive:
  `In Range: 1 | Tip Switch: 0 | Barrel Switch: 0 | Invert: 1 | Eraser: 0` and then
  `In Range: 1 | Tip Switch: 0 | Barrel Switch: 0 | Invert: 0 | Eraser: 1`

But here, the doc says that in the `Tail-end eraser implementation`, we are supposed to keep the `Invert` bit on. There is no mention on how the `button eraser implementation` should behave (so likely like your device is behaving), which probably explains why it works under windows.
The problem is: under Linux, we rely on `Invert` to guess which `BTN_TOOL_*` we need to send, and if the firmware releases it, then we consider this as a `BTN_TOOL_PEN` which explains the whole mess and inability to use the eraser side.

One thing to note (not sure if this should be implemented in the kernel or in libinput) in this particular section of the link above:
> ###  Special Notes for Eraser Button Implementations
>
>  Unlike tail-end eraser implementations, button-based implementations can physically allow the user to activate/deactivate the erase affordance without the pen transitioning through the "out of range" state.   However, this is not supported by the underlying protocol.
>
>  It is highly recommended that, while a pen with an eraser button is in contact with the screen, the eraser switch state be persisted until the pen is lifted, regardless of whether the button is pressed or released. Accidental eraser button presses during the "Pen is in contact" state, and accidental eraser button releases during the "Pen is erasing" state, are common occurrences for users, and the resulting transitions via the "Pen is out of range" state can result in a very jarring user experience.

Basically, when the user starts a `Pen is erasing`, the software should not leave this state unless the `BTN_TOUCH` is set to false (or pressure to null).

**Comment by ShapeShifter499** [comment]
@bentiss So are both ```Invert: 1 | Eraser: 1```  supposed to be set for it to work as intended with the driver as is now?

**Comment by bentiss** [comment]
> @bentiss So are both `Invert: 1 | Eraser: 1` supposed to be set for it to work as intended with the driver as is now?

It should help, yes. The first kernel issue might get in the way however. The kernel expects `Invert` to be before `In Range` in the report, which messes things up a fair bit. However, I *think* this should only impact the very first event, but tests show everything is messed up, so I don't really know what is going on without doing more debugging

**Comment by ShapeShifter499** [comment]
@bentiss anything else I can do to help right now?

**Comment by bentiss** [comment]
> @bentiss anything else I can do to help right now?

Well, right now, I don't think so.

I must apology in advance, but I am not sure I'll be able to work on this in the very next future. This is the type of bug where you can't really predict how quick they will be fixed because there are some fundamental errors in the processing, and we need to make sure we do not regress while fixing it.

**Comment by ShapeShifter499** [comment]
@bentiss there is a plan to fix this bug at some point right?  I'm most likely to keep this tablet, even if I go out and buy another that's better supported in Linux with built in pen support. So this hardware will be around if you want me to test anything new or provide some more debugging logs.

**Comment by bentiss** [comment]
> @bentiss there is a plan to fix this bug at some point right?

yes, of course. When there is a bug, there is a need for a fix. Especially in this case, where your device is using the generic path. Which means other tablets may have the same issue.

Unfortunately, I can't give you an estimate on when I'll be able to work on that.

But thanks for offering your help in the future :)

**Comment by ShapeShifter499** [comment]
@bentiss
> To sum up, if, when you normally press the side button, the logs do not show `Barrel Switch: 1` that means that the kernel is not seeing the physical button press, and that there is nothing we can do to fix it.

The logs DO show `Barrel Switch: 1`  Am I missing something?

**Comment by ShapeShifter499** [comment]
@bentiss @whot  Ok good news! I have eraser working!  

Bad news, I have a totally new issue. If the pen is tilted while writing in the down position I get "phantom erasing". 

Added the following to a xorg configuration under /etc/X11/xorg.conf.d/60-wacom.conf

```
Section "InputClass"   
        Identifier "Nuvision Solo 10 Draw"   
        MatchProduct "04F3200A:00 04F3:22F7"   
        MatchDevicePath "/dev/input/event*"   
        Driver "wacom"   
        Option "Gesture" "off"   
EndSection  
```

**Comment by whot** [comment]
Do you get those phantom erases with the wacom driver enabled or not? But honestly, given the fw is so busted I wouldn't be surprised if some tilt sets/unsets the eraser bit, causing the driver to get confused.

**Comment by ShapeShifter499** [comment]
@whot what do you mean enabled or not?  Without that above xorg configuration using the built-in eraser does not work at all. No odd switching between writing and erasing while the pen tip is against the screen.

**Comment by ShapeShifter499** [comment]
@whot @bentiss I recently also got a used Microsoft Surface Go. I experience the same issue there. I'm going to open up a new issue here in a few with some debugging logs from the various tools I used here. 

Related to

**Comment by ShapeShifter499** [comment]
@whot @bentiss I tried switching to Wayland as I did for #75. I no longer experience the stylus triggering the eraser. But the eraser end is hit or miss. While running Wayland, trying to erase with the eraser end of my stylus (same stylus as #75) caused Xournal and Stylus Labs Write to crash once but I cannot get it to occur again. I'll try to submit a log if there is another application crash, though that could be because of Wayland.

Most likely the firmware of the tablet causing issues on top of whatever is also causing issues in software.

**Comment by ShapeShifter499** [comment]
@whot @bentiss I'm noticing that unlike the Surface Go logs from #75 that the logs here seem to show that this tablet doesn't sense tilt. That's the only difference I'm seeing right now between the logs of the Surface Go and this tablet Nuvision.

Is this still considered to be a different bug than #75?

**Comment by ShapeShifter499** [comment]
@whot @bentiss appears fixed by the patches that @jigpu has posted. Details at https://github.com/linuxwacom/xf86-input-wacom/issues/52

**Comment by whot** [comment]
help me out here - does this mean we can close this one?

**Comment by ShapeShifter499** [comment]
I have to reinstall the operating system on this tablet, I'll retest with patches and let you know by tomorrow.

**Comment by jigpu** [comment]
Assuming we'll need to merge the tablet definition into libwacom and xorg.conf.d snippet into xf86-input-wacom, but waiting for confirmation.

**Comment by ShapeShifter499** [comment]
@jigpu @whot sorry for the late reply but yes I'm confirming this works after the patches from @jigpu. This also requires the xorg.conf.d snippet. Because I've switched to using XFCE as far as I know nothing really makes use of tablet files. But adding a tablet file for this would be useful for people using gnome and others that support it. 

This can be closed as working after https://github.com/linuxwacom/xf86-input-wacom/pull/62 was added.

**Comment by jigpu** [comment]
@ShapeShifter499 could you open a pull requests for the snippet and (if you have one) tablet file based on whatever changes you've made?

**Comment by ShapeShifter499** [comment]
@jigpu I will tomorrow.  I do not use a tablet file currently... do you know if there is any good reasons to use on in XFCE?

**Comment by jigpu** [comment]
My understanding is that libwacom isn't used by any XFCE components. GNOME is the main user at the moment, with KDE just beginning to dip their toes.
